blogpost/post.md

```diff
@@ -0,0 +1,261 @@
+
+def cuda_fn(X, means):
+    with torch.device("cuda"):
+        return tensor_fn(X, means)
+```
+
+This function now takes tensors in CUDA memory and returns tensors in CUDA
+memory, but the function itself is written in NumPy! When we keep the tensors
+in CUDA and perform the computations in `float32`, we see a **200x speed-up**
+over the initial NumPy implementation on `float32` arrays.
+
+**Mixing NumPy and PyTorch**. In this example, we had to write a small adaptor
+to convert tensors to ndarrays and then back to tensors. In programs that mix PyTorch and
+NumPy this is already done by calling `x.detach().cpu().numpy()` (or simply
+`x.numpy(force=True)`). Since when running under `torch.compile` we can run
+NumPy code in CUDA, we can simply modify this code to call `x.numpy()` and when
+running it under `device("cuda")`, as we did above, it will generate efficient
+CUDA code from original NumPy calls without copying the data from CUDA to CPU
+at all. Note that the resulting code would not run without `torch.compile`. For
+it to run in eager mode one would need to rollback to `x.numpy(force=True)`.
+
+## Further Speed-up tricks
+
+**General advice**. The CUDA code we have shown is already quite efficient, but
+it is true that the running example is rather short. When dealing with larger
+programs, we may need to tweak parts of it to make it more efficient. A good
+place to start is the [`torch.compile` troubleshooting
+page](https://pytorch.org/docs/stable/dynamo/troubleshooting.html#performance-profiling).
+This showcases a number of ways to inspect the tracing process, and how to
+identify problematic code that may cause slowdowns.
+
+**Advice when compiling NumPy code**. NumPy, even if rather similar to
+PyTorch, is often used very differently. It is rather common to perform
+computations in NumPy and then do an if/else depending on values within the
+array, or perform operations in-place, perhaps via boolean masks. These
+constructions, while supported by `torch.compile`, hamper its performance.
+Changes like moving from in-place indexing to using `np.where`, writing the
+code in a branchless way, or avoid using in-place ops in favor of out-of-place
+ops can go a long way.
+
+To write fast NumPy code, it is best to avoid loops, but sometimes they are
+unavoidable. When tracing through a loop, `torch.compile` will try to fully
+unroll it. This is sometimes desirable, but sometimes it may not even be
+possible, like when we have a dynamic stopping condition, like in a while loop.
+In these cases, it may be best to just compile the body of the loop, perhaps
+a few iterations at a time (loop unrolling).
+
+**Debugging NumPy code**. Debugging is rather tricky when a compiler is
+involved. To figure out whether an error you are hitting is a `torch.compile`
+error, or an error from the program, you can execute your NumPy program without
+`torch.compile` by replacing the NumPy import by `import torch._numpy as np`.
```

Comment:
Actual question: is torch.compile(backend="eager") equivalent?

Reply:
Almost!
The flow chart is:
- If it fails with `torch._numpy as np` it means that the error is in the decomposition
- If it doesn't, but fails on `torch.compile(backend="eager")`, the error is in dynamo
- If it doesn't, but fails on `torch.compile`, the error is in inductor

The story in the post does not split the second and third case, as the user doesn't care about where in `torch.compile` the error is.